File: 2024/public/js/koppe.js
/*
 * @package koppefest
 *
 * @file jquery handler
 * @author Christoph Kappel <christoph@unexist.dev>
 * @version $Id: 2024/koppe.rb,v 0 1724858416.0-7200 unexist $
 *
 * This program can be distributed under the terms of the GNU GPLv3.
 * See the file COPYING for details.
 */

$(document).ready(function() {

    /**
     * Show toast message
     *
     * @param {String}    str    Message string to show
     ***/

    function showToast(str) {
        /* Insert flash element */
        var flash = $('<div class="flash" style="display: none;"></div>');
        var box     = $('<div class="box primary"></div>');
        var mesg    = $('<span></span>');

        mesg.append(str);
        box.append(mesg);
        flash.append(box);

        /* Click to hide box immediately */
        flash.click(function() {
            flash.remove();
        });

        /* Find count of flash elements */
        var flashes = $('div.flash');

        $('#page').append(flash);

        if (0 < flashes.length) {
            flash.css('top', (parseInt(flash.css('top'))
                - flashes.length * (flash.outerHeight(true) + 10)) + 'px');
        }

        /* Display message for some time */
        flash.fadeIn('slow').delay(5000).hide('slow', function() {
            var height = flash.outerHeight(true);

            flash.remove();

            var flashes = $('div.flash');

            if (0 < flashes.length) {
                for(var i = 0; i < flashes.length; i++) {
                    var f = $(flashes[i]);

                    f.css('top', (parseInt(f.css('top')) -
                        - height + 10) + 'px');
                }
            }
        });
    }

    /* Handle cancellation */
    $('.cancel').click(function(e) {
        var elems = $('input[type="number"], textarea');

        elems.each(function(idx, elem) {
            var that = $(elem);

            if (that.is('input')) {
                that.val(0);
            } else {
                that.val('');
            }
        });

        $('.submit').trigger('click');
    });


    /* Handle submit */
    $('.submit').click(function(e) {
        const fields = [
            'houseno', 'adultsno', 'childrenno', 'beerbenchsno',
            'gardentablesno', 'chairsno', 'standtablesno', 'walltablesno',
            'tentsno', 'grillsno', 'parasolsno'
        ];
        var errno = 0;
        var data = {};

        /* Sanity check fields */
        $(fields).each(function(idx, elemName) {
            var elem = $('#' + elemName);
            var value = elem.val();

            if ("" === value || isNaN(value)) {
                elem.css('border-color', 'red');
                errno++;
            } else {
                data[elemName] = value;
            }
        });

        if (0 === errno) {
            $.ajax({
                type: 'POST',
                url: '/visit',
                data: $.extend({}, data, {
                    foods: $('#foods').val(),
                    drinks: $('#drinks').val(),
                    misc: $('#misc').val(),
                    buildup: $('#buildup').is(':checked') ? 1 : 0,
                    teardown: $('#teardown').is(':checked') ? 1 : 0,
                }),

                /* Success handler */
                success: function (data, status) {
                    /* Insert data */
                    $('#list').prepend($('<div style="margin-top: 30px"><span class="icon">&sharp;'
                        + $('#houseno').val() + '</span></div>'));

                    /* Reset errors */
                    $(fields).each(function(idx, elemName) {
                        var elem = $('#' + elemName);

                        elem.css('border-color', 'black');
                    });

                    /* Update UI */
                    showToast('Wir freuen uns!');
                },

                /* Error handler */
                error: function (xhr, status, e) {
                    showToast('Das ging schief: ' + xhr.responseText);
                }
            });
        } else {
            showToast("Bitte alle rot umrandeten Felder ausfüllen!");
        }
    });
});

File: 2024/koppe.rb
# -*- encoding: utf-8 -*-
#
# @package koppefest
#
# @file Sinatra handler
# @author Christoph Kappel <christoph@unexist.dev>
# @version $Id$
#
# This program can be distributed under the terms of the GNU GPLv3.
# See the file COPYING for details.
#

require "rubygems"
gem "sinatra", "1.4.8"
require "sinatra"
require "haml"
require "data_mapper"
require "json"

# DatamMpper
DataMapper.setup(:default, "sqlite3://#{Dir.pwd}/koppe.db")

# Config
configure do
  set :port, 11000
  set :server, 'webrick'
end

class Houses
  include DataMapper::Resource

  property :id,             Serial
  property :houseno,        Integer, :unique_index => :index_houseno_u
  property :adultsno,       Integer
  property :childrenno,     Integer
  property :beerbenchsno,   Integer
  property :gardentablesno, Integer
  property :chairsno,       Integer
  property :standtablesno,  Integer
  property :walltablesno,   Integer
  property :tentsno,        Integer
  property :grillsno,       Integer
  property :parasolsno,     Integer
  property :foods,          Text
  property :drinks,         Text
  property :misc,           Text
  property :buildup,        Integer
  property :teardown,       Integer
  property :created_at,     DateTime
end

DataMapper.finalize
Houses.auto_upgrade!

# Routes
get "/" do
  @houses = Houses.all(:order => [ :created_at.desc ])

  @adultsno = 0
  @childrenno = 0
  @beerbenchsno = 0
  @gardentablesno = 0
  @chairsno = 0
  @standtablesno = 0
  @walltablesno = 0
  @tentsno = 0
  @grillsno = 0
  @parasolsno = 0
  @buildup = 0
  @teardown = 0

  @houses.each do |h|
    @adultsno += h.adultsno
    @childrenno += h.childrenno
    @beerbenchsno += h.beerbenchsno
    @gardentablesno += h.gardentablesno
    @chairsno += h.chairsno
    @standtablesno += h.standtablesno
    @walltablesno += h.walltablesno
    @tentsno += h.tentsno
    @grillsno += h.grillsno
    @parasolsno += h.parasolsno
    @buildup += h.buildup
    @teardown += h.teardown
  end

  haml :index
end

post "/visit" do
  begin
    Houses.create(
      houseno: params["houseno"],
      adultsno: params["adultsno"],
      childrenno: params["childrenno"],
      beerbenchsno: params["beerbenchsno"],
      gardentablesno: params["gardentablesno"],
      chairsno: params["chairsno"],
      standtablesno: params["standtablesno"],
      walltablesno:  params["walltablesno"],
      tentsno: params["tentsno"],
      grillsno: params["grillsno"],
      parasolsno: params["parasolsno"],
      foods: params["misc"],
      drinks: params["drinks"],
      misc: params["misc"],
      buildup: params["buildup"],
      teardown: params["teardown"],
      created_at: Time.now
    )
  rescue DataObjects::IntegrityError => err
    halt 400, "Diese Hausnummer wurde schon eingetragen!"
  end

  200
end
